Dockerfile

```diff
@@ -6,6 +6,14 @@ RUN set -ex && \
     apt-get install -y --no-install-recommends \
     python3-pip
 
+# Install Rust cargo.
+RUN set -ex && \
+    apt-get install -y --no-install-recommends \
+        curl \
+        build-essential
+RUN curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | bash -s -- -y
```

Comment:
Isn't there a better way to install the rust toolchain than `curl | bash` as root?

Reply:
We could probably install the Debian/Ubuntu distribution instead, but IMO this will serve us better over time: the system distribution tends to lag behind the latest stable, and the Rust dependency ecosystem (unfortunately?) tends to adopt stabilized features relatively quickly.

I think the risk of doing so is similar to the risk of installing a bunch of Python development dependencies as root, which we do immediately below as well 🙂

Reply:
For better or worse this is the default recommended installation per https://www.rust-lang.org/tools/install. `cargo` is available, but not `rustup` from `apt`. However, the `cargo` version is out of date:


```
this version of Cargo is older than the `2021` edition, and only supports `2015` and `2018` editions.
```

And the only method of updating `cargo` AFAIK is via `rustup`.